doc/mgmt/SONiC_PIM_SM_HLD.md

```diff
@@ -0,0 +1,1036 @@
+3) curl -X PATCH "https://100.104.98.88/restconf/data/openconfig-network-instance:network-instances/network-instance=default/protocols/protocol=PIM,pim/pim/global" -H "accept: */*" -H "Authorization: Basic YWRtaW46YnJvYWRjb20=" -H "Content-Type: application/yang-data+json" -d "{\"openconfig-network-instance:global\":{\"rendezvous-points\":{\"rendezvous-point\":[{\"address\":\"4.4.4.4\",\"config\":{\"address\":\"4.4.4.4\",\"multicast-groups\":\"P1\"}}]}}}"
+4) curl -X GET "https://100.104.98.88/restconf/data/openconfig-network-instance:network-instances/network-instance=default/protocols/protocol=PIM,pim/pim/global" -H "accept: application/yang-data+json" -H "Authorization: Basic YWRtaW46YnJvYWRjb20="
+5) curl -X DELETE "https://100.104.98.88/restconf/data/openconfig-network-instance:network-instances/network-instance=default/protocols/protocol=PIM,pim/pim/global" -H "accept: */*" -H "Authorization: Basic YWRtaW46YnJvYWRjb20="
+
+### 3.5.4 gNMI Support
+
+
+# 4 Flow Diagrams
+
+## 4.1 PIM Config Flow
+![](pim_images/PIM_STATIC_RP_CONFIG_FLOW.png)
+
+## 4.2 PIM Data packet Flow
+![](pim_images/2021-12-06-13-10-56.png)
+
+## 4.3 PIM KAT system wide Flow
+!![](pim_images/2022-03-14-22-18-55.png)
+
+## 4.4 Process restart
+!![](pim_images/2022-03-14-22-37-58.png)
+
+
+# 5  Design Constraints and Risks
+
+Ipmcfpmsyncd is single threaded. Till now it has only single entry point which is kernel netlink events. Now through MPFM channel it would route status get from FRR which also has to be serivced. This channel get query request is blocking call in FRR,  (i.e) FRR will give a get request and it will wait for response from ipmcfpmsynd. If Ipmcfpmsynd is busy handling netlink events, FRR would be blocked for a long time, this will lead to frr protocol events not being serviced and hence will lead to functionality issue. 
+Ipmcfpmsyncd has to be implemented as multi threaded process to handle multiple events... Still coming up with multithreaded design.
+          
+# 6 Error Handling
+# 7 Serviceability and Debug
+FRR PIM has debug tracing support. This feature utilizes the same.
+
+# 8 Warm Boot Support
+
+Administrator can initiate Warmboot when PIM is enabled, but there may be an interruption in IP multicast traffic flow. . PIM clears all the existing the IP multicast route entries from APP_DB.So, PIM router will effectively start as in cold-boot.
+
+# 9 Scalability
+Below are the IP multicast scaling numbers:
+
+###### Table 3: IP Multicast Scaling limits
+| Name                                 | Scaling value |
+|--------------------------------------|---------------|
+| Number of Multicast Route Entries    | 8192          |
+| Number of PIM Neighbours             | 64            |
+
+Linux kernel supports only 32 multicast routing interfaces of which one interface is reserved for PIM register interface.  So, PIM/IGMP can be enabled on only 31 routing interfaces.
```

Comment:
Does it mean, we limit the number of interfaces where pim can be enabled to 31 interfaces only in sonic-cli also?